x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FailShardsOnInvalidLicenseClusterListener.java

```diff
@@ -0,0 +1,94 @@
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.RerouteService;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.shard.IndexEventListener;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.license.LicenseStateListener;
+import org.elasticsearch.license.XPackLicenseState;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class FailShardsOnInvalidLicenseClusterListener implements LicenseStateListener, IndexEventListener {
+
+    private static final Logger logger = LogManager.getLogger(FailShardsOnInvalidLicenseClusterListener.class);
+
+    private final XPackLicenseState xPackLicenseState;
+
+    private final RerouteService rerouteService;
+
+    final Set<IndexShard> shardsToFail = new HashSet<>();
+
+    private boolean allowed;
+
+    public FailShardsOnInvalidLicenseClusterListener(XPackLicenseState xPackLicenseState, RerouteService rerouteService) {
+        this.xPackLicenseState = xPackLicenseState;
+        this.rerouteService = rerouteService;
+        this.allowed = xPackLicenseState.isAllowed(XPackLicenseState.Feature.SEARCHABLE_SNAPSHOTS);
+        xPackLicenseState.addListener(this);
+    }
+
+    @Override
+    public synchronized void afterIndexShardStarted(IndexShard indexShard) {
+        shardsToFail.add(indexShard);
+        failActiveShardsIfNecessary();
+    }
+
+    @Override
+    public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {
+        if (indexShard != null) {
+            shardsToFail.remove(indexShard);
+        }
+    }
+
+    @Override
+    public synchronized void licenseStateChanged() {
+        final boolean allowed = xPackLicenseState.isAllowed(XPackLicenseState.Feature.SEARCHABLE_SNAPSHOTS);
+        if (allowed && this.allowed == false) {
+            rerouteService.reroute("reroute after license activation", Priority.NORMAL, new ActionListener<ClusterState>() {
+                @Override
+                public void onResponse(ClusterState clusterState) {
+                    logger.trace("successful reroute after license activation");
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug("unsuccessful reroute after license activation");
+                }
+            });
+        }
+        this.allowed = allowed;
+        failActiveShardsIfNecessary();
+    }
+
+    private void failActiveShardsIfNecessary() {
+        assert Thread.holdsLock(this);
+        if (allowed == false) {
+            for (IndexShard indexShard : shardsToFail) {
+                try {
+                    indexShard.failShard("invalid license", null);
```

Comment:
`failShard()` does a lot of things, some of them involving IO, and we're doing this under the same mutex as the `IndexEventListener` methods which are called during cluster state updates. Are you sure that's ok?

Reply:
I initially contemplated moving the actual failing to a separate (generic) thread, to avoid cluster state updates from being potentially delayed by this. As the triggering of this should (hopefully) be a rare event, and we do similar work on the cluster state update thread when closing a shard, I thought it might unnecessarily complicate this code here with only little benefit. Let me know what your preference is.

Reply:
Ok, on reflection this way seems ok as it should indeed be rare to call this and we won't be retrying very much either.